### PR TITLE
Update shell container creation.

### PR DESCRIPTION
**Pull request by jtkech**
- Don't group anymore services by service type.

- For a generic type definition which is not itself a type but which is rather used to create other constructed generic types, we just pass the service descriptor as onother non singleton service.

- Then, constructed generic types (singleton or not) are registered as other non generic types.

- When multiple services of the same type, we resolve only the singletons ones from the main container, not the non singleton of this type, so without having to create a intermediate scope. We let the others non singleton of this type be added as the others non singleton services.

### CODE REVIEW

File: src/OrchardCore/OrchardCore/Environment/Shell/Builders/ShellContainerFactory.cs [hunk 1]
```diff
@@ -159,5 +153,23 @@ private void EnsureApplicationFeature()
                 }
             }
         }
+
+        internal Type GetImplementationType(ServiceDescriptor descriptor)
+        {
+            if (descriptor.ImplementationType != null)
```
Comment (Jetski5822) [inline src/OrchardCore/OrchardCore/Environment/Shell/Builders/ShellContainerFactory.cs]: (descriptor.ImplementationType is object)  is faster and better IL :)

Comment (Jetski5822) [reply src/OrchardCore/OrchardCore/Environment/Shell/Builders/ShellContainerFactory.cs]: https://twitter.com/FransBouma/status/1115201371182071808

Comment (jtkech) [reply src/OrchardCore/OrchardCore/Environment/Shell/Builders/ShellContainerFactory.cs]: Really? Thanks for the info, will do

Comment (Jetski5822) [reply src/OrchardCore/OrchardCore/Environment/Shell/Builders/ShellContainerFactory.cs]: hehehe

**Comment by jtkech** [comment]
Maybe here i fixed the intermitent setup issue on functional tests.
I will retry it later (i'm going to work on another branch), if so maybe better to do another PR.

**Comment by sebastienros** [comment]
I thought it was on Blogs ?

**Comment by sebastienros** [comment]
Sorry didn't see the two changes

**Comment by sebastienros** [comment]
what is page.click doing then?

**Comment by jtkech** [comment]
As i remember what i saw yesterday it that we can pass a tiemeout to some metods and if not they use a 30 seconds by default. But page.click () has no timeout option and seems to not wait.

**Comment by jtkech** [comment]
That's why here we wait for the waitForNavigation promise.

**Comment by sebastienros** [comment]
https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pageclickselector-options

you are correct, as documented, please create a PR that fixes it individually, this is a huge fix, thanks

**Comment by jtkech** [comment]
Okay i will do.

**Comment by jtkech** [comment]
I think this PR is ready.

- It also wrap cloned host singleton descriptor so that at the tenant level we can retrieve them and e.g their original implementation type. It is useful when we want to deal with a tenant level service collection e.g to isolate some tenant services from the host.

- I included it for testing in the migration to 3.0 branch, and use it to isolate tenant routing services.

**Pull request merged**